standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -5094,26 +5094,22 @@ public static String verifyStatsChangeCtx(String fullTableName, Map<String, Stri
     if (validWriteIds != null && writeId > 0) {
       return null; // We have txn context.
     }
-    String oldVal = oldP == null ? null : oldP.get(StatsSetupConst.COLUMN_STATS_ACCURATE);
-    String newVal = newP == null ? null : newP.get(StatsSetupConst.COLUMN_STATS_ACCURATE);
-    // We don't need txn context is that stats state is not being changed.
-    if (StringUtils.isEmpty(oldVal) && StringUtils.isEmpty(newVal)) {
+
+    if (!StatsSetupConst.areBasicStatsUptoDate(newP)) {
+      // The validWriteIds can be absent, for example, in case of Impala alter.
```

Comment:
Since some components(such as Impala&Trino) may don't care the field `validWriteIds` , can we set the fileld a default value to fix this issue? Like `engine `field i have done:
https://github.com/apache/hive/blob/0f8190fefa56513a042566ef87b852e8343610aa/standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift#L903-L904

It seems that we have introduced some extra filelds in stats related class since Hive4, but the stats related class are often used by other components(Impala&Trino, etc), so some incompatible issues are reported by the other components(Impala&Trino, etc). In one word, if we can set a default value for these new added field, maybe we can fix  similar issues at its root?

Reply:
won't it cause an issue if we set engine = "hive" for Impala? Despite Impala stats is stored in HMS, it can't be used by Hive.

Reply:
Impala could retset the engine fileld value if it doesn't want to use hive stats.

Reply:
As my understanding, there is no easy way to  set `validWriteIds` to a default value, and if the client is HiveServer2, we should always get a non-null `validWriteIds` on the HMS side.
The main idea of the change is that as long as the stats is marked as invalid(stale) on the new target table, we should just let the alter operation go regardless of whether the `validWriteIds` is null or not, as it won't pollute the Hive stats.

Reply:
can we use for that purpose: `TableStatsRequest.isSetValidWriteIdList()`

Reply:
@dengzhhu653 why do we need to lose validation? Would it work if we drop Impala stats on rename?

Reply:
@deniskuzZ, sorry for the late reply!
The `validWriteIds` from impala engine is null, when the impala alters the acid table(like the Test2 in the description), it would be failed as the message shows:
>
ERROR : Failed
org.apache.hadoop.hive.ql.metadata.HiveException: Unable to alter table. Cannot change stats state for a transactional table default.parqtest without providing the transactional write state for verification (new write ID 6, valid write IDs null; current state null; new state {} 

If `StatsSetupConst.areBasicStatsUptoDate(newP)` is false for the new target table, then the alter operation should go as it has marked the stats as stale in the target table, it won't pollute the stats, nor cause trouble for Hive query optimization.